llama_index/agent/openai_agent.py

```diff
@@ -145,10 +244,64 @@ async def achat(
             response = await self._llm.achat(chat_history, functions=functions)
             ai_message = response.message
             chat_history.append(ai_message)
-            function_call = ai_message.additional_kwargs.get("function_call", None)
+            function_call = self._get_latest_function_call(chat_history)
 
         return Response(ai_message.content)
 
+    def astream_chat(
+        self, message: str, chat_history: Optional[List[ChatMessage]] = None
+    ) -> AsyncGenerator[StreamingChatResponse, None]:
+        chat_history = chat_history or self._chat_history
+        tools, functions = self._init_chat(chat_history, message)
+
+        async def _astream_chat(
+            chat_history: List[ChatMessage],
+        ) -> AsyncGenerator[StreamingChatResponse, None]:
+            # TODO: Support forced function call
+            chat_stream_response = StreamingChatResponse(
+                await self._llm.astream_chat(chat_history, functions=functions)
+            )
+
+            # Get the response in a separate thread so we can yield the response
+            thread = Thread(
+                target=chat_stream_response.write_response_to_history,
+                args=(chat_history,),
+            )
+            thread.start()
+            yield chat_stream_response
+            thread.join()
+
+            n_function_calls = 0
+            function_call = self._get_latest_function_call(chat_history)
+            while function_call is not None:
+                if n_function_calls >= self._max_function_calls:
+                    print(f"Exceeded max function calls: {self._max_function_calls}.")
+                    break
+
+                function_message = call_function(
+                    tools, function_call, verbose=self._verbose
+                )
+                chat_history.append(function_message)
+                n_function_calls += 1
+
+                # send function call & output back to get another response
+                chat_stream_response = StreamingChatResponse(
```

Comment:
surprised that typing works out here, since the `astream_chat` returns an async iterato.

Reply:
I think I need to merge main again lol